File: libs/fleet/sdk-bindings/examples/ruby/app_controlled.rb
require 'json'
require 'cyclops_sdk'

require 'json'
require 'thread'
Expected = Struct.new(:method, :url, :headers, :body, :status, :response)
BASE = 'https://cyclops.invalid'; TOKEN = 'https://keycloak.invalid/token'; JSON_HEADERS = [['accept','application/json'],['content-type','application/json'],['authorization','Bearer offline-token']]
def pool_json; { apiVersion:'cua.ai/v1',kind:'OSGymWorkspacePool',metadata:{namespace:'default',name:'default',labels:nil},spec:{replicas:1,template:{containerDiskImage:'registry.example/desktop:offline'},services:[{name:'mcp',targetPort:8080}]},status:nil }; end
def claim_json(bound=false); value={apiVersion:'osgym.cua.ai/v1alpha1',kind:'OSGymSandboxClaim',metadata:{namespace:'default',name:'default',labels:nil},spec:{sandboxTemplateRef:{name:'default'}},status:nil}; value[:status]={phase:'Bound',sandbox:{name:'offline-sandbox'}} if bound; value; end
def queue
 pool_url="#{BASE}/api/k8s/apis/cua.ai/v1/namespaces/default/osgymworkspacepools/default"; claim_url="#{BASE}/api/k8s/apis/osgym.cua.ai/v1alpha1/namespaces/default/osgymsandboxclaims/default"; pool_body=JSON.generate(pool_json).b; claim_body='{"apiVersion":"osgym.cua.ai/v1alpha1","kind":"OSGymSandboxClaim","metadata":{"namespace":"default","name":"claim-1","labels":null},"spec":{"sandboxTemplateRef":{"name":"default"}},"status":null}'.b
 [Expected.new('POST',TOKEN,[['accept','application/json'],['content-type','application/x-www-form-urlencoded']], 'grant_type=client_credentials&client_id=client-id&client_secret=client-secret'.b,200,{access_token:'offline-token',expires_in:3600}),Expected.new('POST',"#{BASE}/api/namespaces",JSON_HEADERS,'{"name":"default"}'.b,201,{}),Expected.new('POST',pool_url.sub(%r{/default\z}, ''),JSON_HEADERS,pool_body,201,pool_json),Expected.new('POST',claim_url.sub(%r{/default\z}, ''),JSON_HEADERS,claim_body,201,claim_json),Expected.new('GET',claim_url,JSON_HEADERS,nil,200,claim_json(true)),Expected.new('GET',pool_url,JSON_HEADERS,nil,200,pool_json),Expected.new('POST',"#{BASE}/api/svc/default/offline-sandbox-mcp/mcp",[['authorization','Bearer offline-token']],'{"offline":true}'.b,202,'offline service accepted'),Expected.new('DELETE',claim_url,JSON_HEADERS,nil,204,''),Expected.new('DELETE',pool_url,JSON_HEADERS,nil,204,''),Expected.new('DELETE',"#{BASE}/api/namespaces/default",JSON_HEADERS,nil,204,'')]
end

class ScriptedHttpClient < CyclopsSdk::HttpClient
  def initialize; @expected=queue; @mutex=Mutex.new; end
  def execute(request)
    @mutex.synchronize do
      item=@expected.shift or raise 'unexpected request'
      actual=[request.method,request.url,request.headers.map{|h|[h.name,h.value]},request.body]
      raise "request mismatch: #{actual.inspect}" unless actual == [item.method,item.url,item.headers,item.body]
      body=item.response.is_a?(String) ? item.response.b : JSON.generate(item.response).b
      CyclopsSdk::HttpResponse.new(status:item.status,headers:[],body:body)
    end
  end
  def assert_exhausted!; raise @expected.inspect unless @expected.empty?; end
end

vm_template = CyclopsSdk::VmTemplate.new(container_disk_image: 'registry.example/desktop:offline', command: nil, runtime: nil, runtime_class_name: nil, node_selector: nil, tolerations: nil, image_pull_policy: nil, image_pull_secret: nil, cpu_cores: nil, memory: nil, firmware: nil, probes: nil, services: nil, oidc: nil)
spec = CyclopsSdk::PoolSpec.new(replicas: 1, template: CyclopsSdk::PoolTemplate.new(runtime: nil, runtime_class_name: nil, node_selector: nil, tolerations: nil, command: nil, container_disk_image: vm_template.container_disk_image, image_pull_secret: nil, cpu_cores: nil, memory: nil, firmware: nil, probes: nil, oidc: nil), autoscaling: nil, services: [CyclopsSdk::SandboxService.new(name: 'mcp', target_port: 8080, protocol: nil)])
transport = ScriptedHttpClient.new
credentials = CyclopsSdk::CyclopsCredentials.new('client-id', 'client-secret')
client = CyclopsSdk::CyclopsClient.connect(CyclopsSdk::CyclopsConfiguration.new(base_url: 'https://cyclops.invalid', token_url: 'https://keycloak.invalid/token', credentials: credentials, pool_poll_interval_ms: 1, pool_poll_limit: 1, claim_poll_interval_ms: 1, claim_poll_limit: 2), transport)
pool = client.create_pool(CyclopsSdk::CreatePoolRequest.new(namespace: 'default', spec: spec))
claim = client.create_claim(CyclopsSdk::CreateClaimRequest.new(pool: pool, spec: CyclopsSdk::ClaimSpec.new(sandbox_template_ref: CyclopsSdk::SandboxTemplateRef.new(name: pool.metadata.name), warmpool: nil, bind_deadline: nil, lifecycle: nil)))
sandbox = client.wait_claim(claim)
service = client.service_request(sandbox, 'mcp', '/mcp', CyclopsSdk::HttpRequest.new(method: 'POST', url: 'https://ignored.invalid/mcp', headers: [], body: '{"offline":true}'.b))
client.delete_claim(claim)
client.delete_pool(pool)
transport.assert_exhausted!
raise 'unexpected service response' unless service.status == 202
puts "pool=#{pool.metadata.name} claim=#{claim.metadata.name} sandbox=#{sandbox.name} service_status=#{service.status}"

File: libs/fleet/sdk-bindings/examples/swift/AppControlled.swift
import Foundation

actor ScriptedHttpClient: HttpClient {
    private var expected: [(String, String, UInt16, String)] = [
        ("POST", "https://keycloak.invalid/token", 200, #"{"access_token":"offline-token","expires_in":3600}"#),
        ("POST", "https://cyclops.invalid/api/namespaces", 201, "{}"),
        ("POST", "https://cyclops.invalid/api/k8s/apis/osgym.cua.ai/v1alpha1/namespaces/default/osgymsandboxwarmpools", 201, #"{"apiVersion":"osgym.cua.ai/v1alpha1","kind":"OSGymSandboxWarmPool","metadata":{"namespace":"default","name":"default","labels":null},"spec":{"replicas":1,"sandboxTemplateRef":{"name":"default"}},"status":null}"#),
        ("POST", "https://cyclops.invalid/api/k8s/apis/osgym.cua.ai/v1alpha1/namespaces/default/osgymsandboxclaims", 201, #"{"apiVersion":"osgym.cua.ai/v1alpha1","kind":"OSGymSandboxClaim","metadata":{"namespace":"default","name":"default","labels":null},"spec":{"sandboxTemplateRef":{"name":"default"}},"status":{"phase":"Bound","sandbox":{"name":"offline-sandbox"}}}"#),
        ("GET", "https://cyclops.invalid/api/k8s/apis/osgym.cua.ai/v1alpha1/namespaces/default/osgymsandboxclaims/default", 200, #"{"apiVersion":"osgym.cua.ai/v1alpha1","kind":"OSGymSandboxClaim","metadata":{"namespace":"default","name":"default","labels":null},"spec":{"sandboxTemplateRef":{"name":"default"}},"status":{"phase":"Bound","sandbox":{"name":"offline-sandbox"}}}"#),
        ("GET", "https://cyclops.invalid/api/k8s/apis/osgym.cua.ai/v1alpha1/namespaces/default/osgymsandboxtemplates/default", 200, #"{"apiVersion":"osgym.cua.ai/v1alpha1","kind":"OSGymSandboxTemplate","metadata":{"namespace":"default","name":"default","labels":null},"spec":{"vmTemplate":{"containerDiskImage":"registry.example/desktop:offline","services":[{"name":"mcp","targetPort":8080}]}}}"#),
        ("POST", "https://cyclops.invalid/api/svc/default/offline-sandbox-mcp/mcp", 202, "offline service accepted"),
        ("DELETE", "https://cyclops.invalid/api/k8s/apis/osgym.cua.ai/v1alpha1/namespaces/default/osgymsandboxclaims/default", 204, ""),
        ("DELETE", "https://cyclops.invalid/api/k8s/apis/osgym.cua.ai/v1alpha1/namespaces/default/osgymsandboxwarmpools/default", 204, ""),
        ("DELETE", "https://cyclops.invalid/api/namespaces/default", 204, "")]
    private var requests = [HttpRequest]()
    func execute(request: HttpRequest) async throws -> HttpResponse { let item = expected.removeFirst(); precondition(request.method == item.0 && request.url == item.1); requests.append(request); return HttpResponse(status: item.2, headers: [], body: Data(item.3.utf8)) }
    func assertExhausted() { precondition(expected.isEmpty) }
    func methods() -> [String] { requests.map(\.method) }
}

@main struct AppControlled {
    static func main() async throws {
        let spec = OsGymSandboxWarmPoolSpec(replicas: 1, sandboxTemplateRef: SandboxTemplateRef(name: "default"), autoscaling: nil)
        let transport = ScriptedHttpClient()
        let credentials = CyclopsCredentials(clientId: "client-id", clientSecret: "client-secret")
        let client = try CyclopsClient.connect(configuration: CyclopsConfiguration(baseUrl: "https://cyclops.invalid", tokenUrl: "https://keycloak.invalid/token", credentials: credentials, poolPollIntervalMs: 1, poolPollLimit: 1, claimPollIntervalMs: 1, claimPollLimit: 2), httpClient: transport)
        let pool = try await client.createPool(request: CreatePoolRequest(namespace: "default", spec: spec))
        let claim = try await client.createClaim(request: CreateClaimRequest(pool: pool, spec: ClaimSpec(sandboxTemplateRef: SandboxTemplateRef(name: pool.metadata.name), warmpool: nil, bindDeadline: nil, lifecycle: nil)))
        let sandbox = try await client.waitClaim(claim: claim)
        let service = try await client.serviceRequest(sandbox: sandbox, service: "mcp", path: "/mcp", request: HttpRequest(method: "POST", url: "https://ignored.invalid/mcp", headers: [], body: Data("{\"offline\":true}".utf8)))
        try await client.deleteClaim(claim: claim)
        try await client.deletePool(pool: pool)
        await transport.assertExhausted()
        let methods = await transport.methods()
        precondition(methods == ["POST", "POST", "POST", "POST", "GET", "GET", "POST", "DELETE", "DELETE", "DELETE"])
        print("pool=\(pool.metadata.name) claim=\(claim.metadata.name) sandbox=\(sandbox.name) service_status=\(service.status)")
    }
}
